No method in multimethod 'type->metadata-writer' for dispatch value: :set

I've started following along with the https://github.com/xtdb/sakila-playground examples in https://www.youtube.com/watch?v=xB-hIpQxweI&t=29s .
Then I've left my REPL running for a few days and why I got back I've seen this error dumped into my CIDER REPL twice:

```
17:27:36 | INFO  xtdb.stagnant-log-flusher | last chunk tx-id null, flushing any pending writes
17:27:36 | ERROR xtdb.util | Uncaught exception:
java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: No method in multimethod 'type->metadata-writer' for dispatch value: :set
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at clojure.core$deref_future.invokeStatic(core.clj:2317)
	at clojure.core$deref.invokeStatic(core.clj:2337)
	at clojure.core$deref.invoke(core.clj:2323)
	at xtdb.indexer.live_index.LiveIndex.finishChunk(live_index.clj:307)
	at xtdb.indexer.Indexer.finish_chunk_BANG_(indexer.clj:716)
	at xtdb.indexer.Indexer.forceFlush(indexer.clj:667)
	at xtdb.log.watcher$watch_log_BANG_$reify__25891.acceptRecord(watcher.clj:61)
	at xtdb.log$tx_handler$fn__23546.invoke(log.clj:35)
	at clojure.core.protocols$iter_reduce.invokeStatic(protocols.clj:49)
	at clojure.core.protocols$fn__8230.invokeStatic(protocols.clj:75)
	at clojure.core.protocols$fn__8230.invoke(protocols.clj:75)
	at clojure.core.protocols$fn__8178$G__8173__8191.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6886)
	at clojure.core$reduce.invoke(core.clj:6868)
	at xtdb.log.NotifyingSubscriberHandler$fn__23579.invoke(log.clj:91)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalArgumentException: No method in multimethod 'type->metadata-writer' for dispatch value: :set
	at clojure.lang.MultiFn.getFn(MultiFn.java:156)
	at clojure.lang.MultiFn.invoke(MultiFn.java:238)
	at xtdb.metadata$__GT_page_meta_wtr$__GT_nested_meta_writer$reify__15277.apply(metadata.clj:239)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1220)
	at xtdb.metadata$__GT_page_meta_wtr$__GT_nested_meta_writer__15275.invoke(metadata.clj:235)
	at xtdb.metadata$__GT_page_meta_wtr$write_col_meta_BANG___15281$fn__15282.invoke(metadata.clj:246)
	at clojure.core$mapv$fn__8535.invoke(core.clj:6979)
	at clojure.core.protocols$iter_reduce.invokeStatic(protocols.clj:49)
	at clojure.core.protocols$fn__8230.invokeStatic(protocols.clj:75)
	at clojure.core.protocols$fn__8230.invoke(protocols.clj:75)
	at clojure.core.protocols$fn__8178$G__8173__8191.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6886)
	at clojure.core$mapv.invokeStatic(core.clj:6970)
	at clojure.core$mapv.invoke(core.clj:6970)
	at xtdb.metadata$__GT_page_meta_wtr$write_col_meta_BANG___15281.invoke(metadata.clj:245)
	at xtdb.metadata$$reify__15293.writeMetadata(metadata.clj:272)
	at xtdb.trie$open_trie_writer$reify__15398.writeLeaf(trie.clj:124)
	at xtdb.trie$write_live_trie_node$write_node_BANG___15416.invoke(trie.clj:190)
	at xtdb.trie$write_live_trie_node.invokeStatic(trie.clj:192)
	at xtdb.trie$write_live_trie_node.invoke(trie.clj:169)
	at xtdb.trie$write_live_trie_BANG_$fn__15437.invoke(trie.clj:203)
	at xtdb.trie$write_live_trie_BANG_.invokeStatic(trie.clj:197)
	at xtdb.trie$write_live_trie_BANG_.invoke(trie.clj:194)
	at xtdb.indexer.live_index.LiveTable$fn__15634.invoke(live_index.clj:183)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
17:27:45 | INFO  xtdb.stagnant-log-flusher | last chunk tx-id null, flushing any pending writes
17:27:45 | ERROR xtdb.util | Uncaught exception:
java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: No method in multimethod 'type->metadata-writer' for dispatch value: :set
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at clojure.core$deref_future.invokeStatic(core.clj:2317)
	at clojure.core$deref.invokeStatic(core.clj:2337)
	at clojure.core$deref.invoke(core.clj:2323)
	at xtdb.indexer.live_index.LiveIndex.finishChunk(live_index.clj:307)
	at xtdb.indexer.Indexer.finish_chunk_BANG_(indexer.clj:716)
	at xtdb.indexer.Indexer.forceFlush(indexer.clj:667)
	at xtdb.log.watcher$watch_log_BANG_$reify__25891.acceptRecord(watcher.clj:61)
	at xtdb.log$tx_handler$fn__23546.invoke(log.clj:35)
	at clojure.core.protocols$iter_reduce.invokeStatic(protocols.clj:49)
	at clojure.core.protocols$fn__8230.invokeStatic(protocols.clj:75)
	at clojure.core.protocols$fn__8230.invoke(protocols.clj:75)
	at clojure.core.protocols$fn__8178$G__8173__8191.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6886)
	at clojure.core$reduce.invoke(core.clj:6868)
	at xtdb.log.NotifyingSubscriberHandler$fn__23579.invoke(log.clj:91)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalArgumentException: No method in multimethod 'type->metadata-writer' for dispatch value: :set
	at clojure.lang.MultiFn.getFn(MultiFn.java:156)
	at clojure.lang.MultiFn.invoke(MultiFn.java:238)
	at xtdb.metadata$__GT_page_meta_wtr$__GT_nested_meta_writer$reify__15277.apply(metadata.clj:239)
<the rest of the stack trace is the same as the previous exception>
```

--- a/core/src/main/clojure/xtdb/metadata.clj
+++ b/core/src/main/clojure/xtdb/metadata.clj
@@ -195,6 +195,19 @@
             (writeContentMetadata [_]
               (.writeInt list-type-wtr data-meta-idx))))))))
 
+(defmethod type->metadata-writer :set [write-col-meta! ^IVectorWriter types-wtr col-type]
+  (let [types-wp (.writerPosition types-wtr)
+        set-type-wtr (.structKeyWriter types-wtr (types/col-type->field-name col-type)
+                                        (FieldType/nullable #xt.arrow/type :i32))]
+    (reify NestedMetadataWriter
+      (appendNestedMetadata [_ content-col]
+        (write-col-meta! (.listElementReader ^IVectorReader content-col))
+
+        (let [data-meta-idx (dec (.getPosition types-wp))]
+          (reify ContentMetadataWriter
+            (writeContentMetadata [_]
+              (.writeInt set-type-wtr data-meta-idx))))))))
+
 (defmethod type->metadata-writer :struct [write-col-meta! ^IVectorWriter types-wtr, col-type]
   (let [types-wp (.writerPosition types-wtr)
         struct-type-wtr (.structKeyWriter types-wtr

--- a/src/test/clojure/xtdb/metadata_test.clj
+++ b/src/test/clojure/xtdb/metadata_test.clj
@@ -1,9 +1,11 @@
 (ns xtdb.metadata-test
-  (:require [clojure.test :as t :refer [deftest]]
+  (:require [clojure.java.io :as io]
+            [clojure.test :as t :refer [deftest]]
             [xtdb.api :as xt]
             [xtdb.expression.metadata :as expr.meta]
             [xtdb.metadata :as meta]
             [xtdb.node :as xtn]
+            [xtdb.test-json :as tj]
             [xtdb.test-util :as tu]
             [xtdb.trie :as trie]
             [xtdb.util :as util])
@@ -154,3 +156,26 @@
                  (.columnNames table-metadata)))
 
         (t/is (true? (.test page-idx-pred 0)))))))
+
+(t/deftest test-set-metadata
+  (let [node-dir (util/->path "target/test-set-metadata")]
+    (util/delete-dir node-dir)
+    (util/with-open [node (tu/->local-node {:node-dir node-dir})]
+
+      (xt/submit-tx node [[:put-docs :xt_docs {:xt/id "foo" :colours #{"red" "blue" "green"}}]])
+
+      (tu/finish-chunk! node)
+
+      (let [^IMetadataManager metadata-mgr (tu/component node ::meta/metadata-manager)
+            ^IBufferPool buffer-pool (tu/component node :xtdb/buffer-pool)
+            meta-file-path (trie/->table-meta-file-path (util/->path "tables/xt_docs/") (trie/->log-trie-key 0 2 1))]
+
+        (util/with-open [{meta-rel-rdr :rdr} (trie/open-meta-file buffer-pool meta-file-path)]
+          (let [table-metadata (.tableMetadata metadata-mgr meta-rel-rdr meta-file-path)]
+
+            (tj/check-json (.toPath (io/as-file (io/resource "xtdb/metadata-test/set")))
+
+                           (.resolve node-dir "objects/v01/tables/"))
+
+            (t/is (= #{"xt$iid" "xt$id" "xt$system_from" "colours" "$data$"}
+                     (.columnNames table-metadata)))))))))

--- a/src/test/resources/xtdb/metadata-test/set/xt$txs/data/log-l00-nr02-rs1.arrow.json
+++ b/src/test/resources/xtdb/metadata-test/set/xt$txs/data/log-l00-nr02-rs1.arrow.json
@@ -1,0 +1,171 @@
+{
+  "schema" : {
+    "fields" : [{
+      "name" : "xt$iid",
+      "nullable" : false,
+      "type" : {
+        "name" : "fixedsizebinary",
+        "byteWidth" : 16
+      },
+      "children" : [ ]
+    },{
+      "name" : "xt$system_from",
+      "nullable" : false,
+      "type" : {
+        "name" : "timestamp",
+        "unit" : "MICROSECOND",
+        "timezone" : "UTC"
+      },
+      "children" : [ ]
+    },{
+      "name" : "xt$valid_from",
+      "nullable" : false,
+      "type" : {
+        "name" : "timestamp",
+        "unit" : "MICROSECOND",
+        "timezone" : "UTC"
+      },
+      "children" : [ ]
+    },{
+      "name" : "xt$valid_to",
+      "nullable" : false,
+      "type" : {
+        "name" : "timestamp",
+        "unit" : "MICROSECOND",
+        "timezone" : "UTC"
+      },
+      "children" : [ ]
+    },{
+      "name" : "op",
+      "nullable" : false,
+      "type" : {
+        "name" : "union",
+        "mode" : "Dense",
+        "typeIds" : [0,1,2]
+      },
+      "children" : [{
+        "name" : "put",
+        "nullable" : false,
+        "type" : {
+          "name" : "struct"
+        },
+        "children" : [{
+          "name" : "xt$id",
+          "nullable" : false,
+          "type" : {
+            "name" : "int",
+            "bitWidth" : 64,
+            "isSigned" : true
+          },
+          "children" : [ ]
+        },{
+          "name" : "xt$tx_time",
+          "nullable" : false,
+          "type" : {
+            "name" : "timestamp",
+            "unit" : "MICROSECOND",
+            "timezone" : "UTC"
+          },
+          "children" : [ ]
+        },{
+          "name" : "xt$committed?",
+          "nullable" : false,
+          "type" : {
+            "name" : "bool"
+          },
+          "children" : [ ]
+        },{
+          "name" : "xt$error",
+          "nullable" : true,
+          "type" : {
+            "name" : "TransitType"
+          },
+          "children" : [ ],
+          "metadata" : [{
+            "value" : "xt/transit+msgpack",
+            "key" : "ARROW:extension:name"
+          },{
+            "value" : "",
+            "key" : "ARROW:extension:metadata"
+          }]
+        }]
+      },{
+        "name" : "delete",
+        "nullable" : true,
+        "type" : {
+          "name" : "null"
+        },
+        "children" : [ ]
+      },{
+        "name" : "erase",
+        "nullable" : true,
+        "type" : {
+          "name" : "null"
+        },
+        "children" : [ ]
+      }]
+    }]
+  },
+  "batches" : [{
+    "count" : 1,
+    "columns" : [{
+      "name" : "xt$iid",
+      "count" : 1,
+      "VALIDITY" : [1],
+      "DATA" : ["a4e167a76a05add8a8654c169b07b044"]
+    },{
+      "name" : "xt$system_from",
+      "count" : 1,
+      "VALIDITY" : [1],
+      "DATA" : [1577836800000000]
+    },{
+      "name" : "xt$valid_from",
+      "count" : 1,
+      "VALIDITY" : [1],
+      "DATA" : [1577836800000000]
+    },{
+      "name" : "xt$valid_to",
+      "count" : 1,
+      "VALIDITY" : [1],
+      "DATA" : [9223372036854775807]
+    },{
+      "name" : "op",
+      "count" : 1,
+      "TYPE_ID" : [0],
+      "OFFSET" : [0],
+      "children" : [{
+        "name" : "put",
+        "count" : 1,
+        "VALIDITY" : [1],
+        "children" : [{
+          "name" : "xt$id",
+          "count" : 1,
+          "VALIDITY" : [1],
+          "DATA" : ["0"]
+        },{
+          "name" : "xt$tx_time",
+          "count" : 1,
+          "VALIDITY" : [1],
+          "DATA" : [1577836800000000]
+        },{
+          "name" : "xt$committed?",
+          "count" : 1,
+          "VALIDITY" : [1],
+          "DATA" : [1]
+        },{
+          "name" : "xt$error",
+          "count" : 1,
+          "VALIDITY" : [0],
+          "OFFSET" : [0,0],
+          "DATA" : [""]
+        }]
+      },{
+        "name" : "delete",
+        "count" : 0
+      },{
+        "name" : "erase",
+        "count" : 0
+      }]
+    }]
+  }]
+}

--- a/src/test/resources/xtdb/metadata-test/set/xt$txs/meta/log-l00-nr02-rs1.arrow.json
+++ b/src/test/resources/xtdb/metadata-test/set/xt$txs/meta/log-l00-nr02-rs1.arrow.json
@@ -1,0 +1,346 @@
+{
+  "schema" : {
+    "fields" : [{
+      "name" : "nodes",
+      "nullable" : false,
+      "type" : {
+        "name" : "union",
+        "mode" : "Dense",
+        "typeIds" : [0,1,2,3]
+      },
+      "children" : [{
+        "name" : "nil",
+        "nullable" : true,
+        "type" : {
+          "name" : "null"
+        },
+        "children" : [ ]
+      },{
+        "name" : "branch-iid",
+        "nullable" : false,
+        "type" : {
+          "name" : "list"
+        },
+        "children" : [{
+          "name" : "$data$",
+          "nullable" : true,
+          "type" : {
+            "name" : "int",
+            "bitWidth" : 32,
+            "isSigned" : true
+          },
+          "children" : [ ]
+        }]
+      },{
+        "name" : "branch-recency",
+        "nullable" : false,
+        "type" : {
+          "name" : "map",
+          "keysSorted" : true
+        },
+        "children" : [{
+          "name" : "entries",
+          "nullable" : false,
+          "type" : {
+            "name" : "struct"
+          },
+          "children" : [{
+            "name" : "recency",
+            "nullable" : false,
+            "type" : {
+              "name" : "timestamp",
+              "unit" : "MICROSECOND",
+              "timezone" : "UTC"
+            },
+            "children" : [ ]
+          },{
+            "name" : "idx",
+            "nullable" : true,
+            "type" : {
+              "name" : "int",
+              "bitWidth" : 32,
+              "isSigned" : true
+            },
+            "children" : [ ]
+          }]
+        }]
+      },{
+        "name" : "leaf",
+        "nullable" : false,
+        "type" : {
+          "name" : "struct"
+        },
+        "children" : [{
+          "name" : "data-page-idx",
+          "nullable" : false,
+          "type" : {
+            "name" : "int",
+            "bitWidth" : 32,
+            "isSigned" : true
+          },
+          "children" : [ ]
+        },{
+          "name" : "columns",
+          "nullable" : false,
+          "type" : {
+            "name" : "list"
+          },
+          "children" : [{
+            "name" : "$data$",
+            "nullable" : false,
+            "type" : {
+              "name" : "struct"
+            },
+            "children" : [{
+              "name" : "col-name",
+              "nullable" : false,
+              "type" : {
+                "name" : "utf8"
+              },
+              "children" : [ ]
+            },{
+              "name" : "root-col?",
+              "nullable" : false,
+              "type" : {
+                "name" : "bool"
+              },
+              "children" : [ ]
+            },{
+              "name" : "count",
+              "nullable" : false,
+              "type" : {
+                "name" : "int",
+                "bitWidth" : 64,
+                "isSigned" : true
+              },
+              "children" : [ ]
+            },{
+              "name" : "types",
+              "nullable" : false,
+              "type" : {
+                "name" : "struct"
+              },
+              "children" : [{
+                "name" : "timestamp-tz-micro-utc",
+                "nullable" : true,
+                "type" : {
+                  "name" : "struct"
+                },
+                "children" : [{
+                  "name" : "min",
+                  "nullable" : true,
+                  "type" : {
+                    "name" : "timestamp",
+                    "unit" : "MICROSECOND",
+                    "timezone" : "UTC"
+                  },
+                  "children" : [ ]
+                },{
+                  "name" : "max",
+                  "nullable" : true,
+                  "type" : {
+                    "name" : "timestamp",
+                    "unit" : "MICROSECOND",
+                    "timezone" : "UTC"
+                  },
+                  "children" : [ ]
+                }]
+              },{
+                "name" : "fixed-size-binary",
+                "nullable" : true,
+                "type" : {
+                  "name" : "bool"
+                },
+                "children" : [ ]
+              },{
+                "name" : "i64",
+                "nullable" : true,
+                "type" : {
+                  "name" : "struct"
+                },
+                "children" : [{
+                  "name" : "min",
+                  "nullable" : true,
+                  "type" : {
+                    "name" : "int",
+                    "bitWidth" : 64,
+                    "isSigned" : true
+                  },
+                  "children" : [ ]
+                },{
+                  "name" : "max",
+                  "nullable" : true,
+                  "type" : {
+                    "name" : "int",
+                    "bitWidth" : 64,
+                    "isSigned" : true
+                  },
+                  "children" : [ ]
+                }]
+              },{
+                "name" : "transit",
+                "nullable" : true,
+                "type" : {
+                  "name" : "bool"
+                },
+                "children" : [ ]
+              },{
+                "name" : "bool",
+                "nullable" : true,
+                "type" : {
+                  "name" : "bool"
+                },
+                "children" : [ ]
+              }]
+            },{
+              "name" : "bloom",
+              "nullable" : true,
+              "type" : {
+                "name" : "binary"
+              },
+              "children" : [ ]
+            }]
+          }]
+        }]
+      }]
+    }]
+  },
+  "batches" : [{
+    "count" : 1,
+    "columns" : [{
+      "name" : "nodes",
+      "count" : 1,
+      "TYPE_ID" : [3],
+      "OFFSET" : [0],
+      "children" : [{
+        "name" : "nil",
+        "count" : 0
+      },{
+        "name" : "branch-iid",
+        "count" : 0,
+        "VALIDITY" : [ ],
+        "OFFSET" : [0],
+        "children" : [{
+          "name" : "$data$",
+          "count" : 0,
+          "VALIDITY" : [ ],
+          "DATA" : [ ]
+        }]
+      },{
+        "name" : "branch-recency",
+        "count" : 0,
+        "VALIDITY" : [ ],
+        "OFFSET" : [0],
+        "children" : [{
+          "name" : "entries",
+          "count" : 0,
+          "VALIDITY" : [ ],
+          "children" : [{
+            "name" : "recency",
+            "count" : 0,
+            "VALIDITY" : [ ],
+            "DATA" : [ ]
+          },{
+            "name" : "idx",
+            "count" : 0,
+            "VALIDITY" : [ ],
+            "DATA" : [ ]
+          }]
+        }]
+      },{
+        "name" : "leaf",
+        "count" : 1,
+        "VALIDITY" : [1],
+        "children" : [{
+          "name" : "data-page-idx",
+          "count" : 1,
+          "VALIDITY" : [1],
+          "DATA" : [0]
+        },{
+          "name" : "columns",
+          "count" : 1,
+          "VALIDITY" : [1],
+          "OFFSET" : [0,6],
+          "children" : [{
+            "name" : "$data$",
+            "count" : 6,
+            "VALIDITY" : [1,1,1,1,1,1],
+            "children" : [{
+              "name" : "col-name",
+              "count" : 6,
+              "VALIDITY" : [1,1,1,1,1,1],
+              "OFFSET" : [0,14,20,25,35,43,56],
+              "DATA" : ["xt$system_from","xt$iid","xt$id","xt$tx_time","xt$error","xt$committed?"]
+            },{
+              "name" : "root-col?",
+              "count" : 6,
+              "VALIDITY" : [1,1,1,1,1,1],
+              "DATA" : [1,1,1,1,1,1]
+            },{
+              "name" : "count",
+              "count" : 6,
+              "VALIDITY" : [1,1,1,1,1,1],
+              "DATA" : ["1","1","1","1","0","1"]
+            },{
+              "name" : "types",
+              "count" : 6,
+              "VALIDITY" : [1,1,1,1,1,1],
+              "children" : [{
+                "name" : "timestamp-tz-micro-utc",
+                "count" : 6,
+                "VALIDITY" : [1,0,0,1,0,0],
+                "children" : [{
+                  "name" : "min",
+                  "count" : 6,
+                  "VALIDITY" : [1,0,0,1,0,0],
+                  "DATA" : [1577836800000000,0,0,1577836800000000,0,0]
+                },{
+                  "name" : "max",
+                  "count" : 6,
+                  "VALIDITY" : [1,0,0,1,0,0],
+                  "DATA" : [1577836800000000,0,0,1577836800000000,0,0]
+                }]
+              },{
+                "name" : "fixed-size-binary",
+                "count" : 6,
+                "VALIDITY" : [0,1,0,0,0,0],
+                "DATA" : [0,1,0,0,0,0]
+              },{
+                "name" : "i64",
+                "count" : 6,
+                "VALIDITY" : [0,0,1,0,0,0],
+                "children" : [{
+                  "name" : "min",
+                  "count" : 6,
+                  "VALIDITY" : [0,0,1,0,0,0],
+                  "DATA" : ["0","0","0","0","0","0"]
+                },{
+                  "name" : "max",
+                  "count" : 6,
+                  "VALIDITY" : [0,0,1,0,0,0],
+                  "DATA" : ["0","0","0","0","0","0"]
+                }]
+              },{
+                "name" : "transit",
+                "count" : 6,
+                "VALIDITY" : [0,0,0,0,1,0],
+                "DATA" : [0,0,0,0,1,0]
+              },{
+                "name" : "bool",
+                "count" : 6,
+                "VALIDITY" : [0,0,0,0,0,1],
+                "DATA" : [0,0,0,0,0,1]
+              }]
+            },{
+              "name" : "bloom",
+              "count" : 6,
+              "VALIDITY" : [1,1,1,1,1,1],
+              "OFFSET" : [0,38,76,114,152,160,182],
+              "DATA" : ["3a30000003000000090000000c0000000f0000002000000022000000240000004c012a6e08db","3a3000000300000004000000080000000f0000002000000022000000240000005b01bd8ef973","3a3000000300000000000000050000000a0000002000000022000000240000000000fc2af855","3a30000003000000090000000c0000000f0000002000000022000000240000004c012a6e08db","3a30000000000000","3a300000010000000000020010000000130026003900"]
+            }]
+          }]
+        }]
+      }]
+    }]
+  }]
+}

--- a/src/test/resources/xtdb/metadata-test/set/xt_docs/data/log-l00-nr02-rs1.arrow.json
+++ b/src/test/resources/xtdb/metadata-test/set/xt_docs/data/log-l00-nr02-rs1.arrow.json
@@ -1,0 +1,202 @@
+{
+  "schema" : {
+    "fields" : [{
+      "name" : "xt$iid",
+      "nullable" : false,
+      "type" : {
+        "name" : "fixedsizebinary",
+        "byteWidth" : 16
+      },
+      "children" : [ ]
+    },{
+      "name" : "xt$system_from",
+      "nullable" : false,
+      "type" : {
+        "name" : "timestamp",
+        "unit" : "MICROSECOND",
+        "timezone" : "UTC"
+      },
+      "children" : [ ]
+    },{
+      "name" : "xt$valid_from",
+      "nullable" : false,
+      "type" : {
+        "name" : "timestamp",
+        "unit" : "MICROSECOND",
+        "timezone" : "UTC"
+      },
+      "children" : [ ]
+    },{
+      "name" : "xt$valid_to",
+      "nullable" : false,
+      "type" : {
+        "name" : "timestamp",
+        "unit" : "MICROSECOND",
+        "timezone" : "UTC"
+      },
+      "children" : [ ]
+    },{
+      "name" : "op",
+      "nullable" : false,
+      "type" : {
+        "name" : "union",
+        "mode" : "Dense",
+        "typeIds" : [0,1,2]
+      },
+      "children" : [{
+        "name" : "put",
+        "nullable" : false,
+        "type" : {
+          "name" : "struct"
+        },
+        "children" : [{
+          "name" : "xt$id",
+          "nullable" : false,
+          "type" : {
+            "name" : "union",
+            "mode" : "Dense",
+            "typeIds" : [ ]
+          },
+          "children" : [{
+            "name" : "utf8",
+            "nullable" : false,
+            "type" : {
+              "name" : "utf8"
+            },
+            "children" : [ ]
+          }]
+        },{
+          "name" : "colours",
+          "nullable" : false,
+          "type" : {
+            "name" : "union",
+            "mode" : "Dense",
+            "typeIds" : [ ]
+          },
+          "children" : [{
+            "name" : "set",
+            "nullable" : false,
+            "type" : {
+              "name" : "SetType"
+            },
+            "children" : [{
+              "name" : "$data$",
+              "nullable" : false,
+              "type" : {
+                "name" : "union",
+                "mode" : "Dense",
+                "typeIds" : [ ]
+              },
+              "children" : [{
+                "name" : "utf8",
+                "nullable" : false,
+                "type" : {
+                  "name" : "utf8"
+                },
+                "children" : [ ]
+              }]
+            }],
+            "metadata" : [{
+              "value" : "set",
+              "key" : "ARROW:extension:name"
+            },{
+              "value" : "",
+              "key" : "ARROW:extension:metadata"
+            }]
+          }]
+        }]
+      },{
+        "name" : "delete",
+        "nullable" : true,
+        "type" : {
+          "name" : "null"
+        },
+        "children" : [ ]
+      },{
+        "name" : "erase",
+        "nullable" : true,
+        "type" : {
+          "name" : "null"
+        },
+        "children" : [ ]
+      }]
+    }]
+  },
+  "batches" : [{
+    "count" : 1,
+    "columns" : [{
+      "name" : "xt$iid",
+      "count" : 1,
+      "VALIDITY" : [1],
+      "DATA" : ["d9c7fae2a04e047164936265ba33cf80"]
+    },{
+      "name" : "xt$system_from",
+      "count" : 1,
+      "VALIDITY" : [1],
+      "DATA" : [1577836800000000]
+    },{
+      "name" : "xt$valid_from",
+      "count" : 1,
+      "VALIDITY" : [1],
+      "DATA" : [1577836800000000]
+    },{
+      "name" : "xt$valid_to",
+      "count" : 1,
+      "VALIDITY" : [1],
+      "DATA" : [9223372036854775807]
+    },{
+      "name" : "op",
+      "count" : 1,
+      "TYPE_ID" : [0],
+      "OFFSET" : [0],
+      "children" : [{
+        "name" : "put",
+        "count" : 1,
+        "VALIDITY" : [1],
+        "children" : [{
+          "name" : "xt$id",
+          "count" : 1,
+          "TYPE_ID" : [0],
+          "OFFSET" : [0],
+          "children" : [{
+            "name" : "utf8",
+            "count" : 1,
+            "VALIDITY" : [1],
+            "OFFSET" : [0,3],
+            "DATA" : ["foo"]
+          }]
+        },{
+          "name" : "colours",
+          "count" : 1,
+          "TYPE_ID" : [0],
+          "OFFSET" : [0],
+          "children" : [{
+            "name" : "set",
+            "count" : 1,
+            "VALIDITY" : [1],
+            "OFFSET" : [0,3],
+            "children" : [{
+              "name" : "$data$",
+              "count" : 3,
+              "TYPE_ID" : [0,0,0],
+              "OFFSET" : [0,1,2],
+              "children" : [{
+                "name" : "utf8",
+                "count" : 3,
+                "VALIDITY" : [1,1,1],
+                "OFFSET" : [0,4,9,12],
+                "DATA" : ["blue","green","red"]
+              }]
+            }]
+          }]
+        }]
+      },{
+        "name" : "delete",
+        "count" : 0
+      },{
+        "name" : "erase",
+        "count" : 0
+      }]
+    }]
+  }]
+}

--- a/src/test/resources/xtdb/metadata-test/set/xt_docs/meta/log-l00-nr02-rs1.arrow.json
+++ b/src/test/resources/xtdb/metadata-test/set/xt_docs/meta/log-l00-nr02-rs1.arrow.json
@@ -1,0 +1,334 @@
+{
+  "schema" : {
+    "fields" : [{
+      "name" : "nodes",
+      "nullable" : false,
+      "type" : {
+        "name" : "union",
+        "mode" : "Dense",
+        "typeIds" : [0,1,2,3]
+      },
+      "children" : [{
+        "name" : "nil",
+        "nullable" : true,
+        "type" : {
+          "name" : "null"
+        },
+        "children" : [ ]
+      },{
+        "name" : "branch-iid",
+        "nullable" : false,
+        "type" : {
+          "name" : "list"
+        },
+        "children" : [{
+          "name" : "$data$",
+          "nullable" : true,
+          "type" : {
+            "name" : "int",
+            "bitWidth" : 32,
+            "isSigned" : true
+          },
+          "children" : [ ]
+        }]
+      },{
+        "name" : "branch-recency",
+        "nullable" : false,
+        "type" : {
+          "name" : "map",
+          "keysSorted" : true
+        },
+        "children" : [{
+          "name" : "entries",
+          "nullable" : false,
+          "type" : {
+            "name" : "struct"
+          },
+          "children" : [{
+            "name" : "recency",
+            "nullable" : false,
+            "type" : {
+              "name" : "timestamp",
+              "unit" : "MICROSECOND",
+              "timezone" : "UTC"
+            },
+            "children" : [ ]
+          },{
+            "name" : "idx",
+            "nullable" : true,
+            "type" : {
+              "name" : "int",
+              "bitWidth" : 32,
+              "isSigned" : true
+            },
+            "children" : [ ]
+          }]
+        }]
+      },{
+        "name" : "leaf",
+        "nullable" : false,
+        "type" : {
+          "name" : "struct"
+        },
+        "children" : [{
+          "name" : "data-page-idx",
+          "nullable" : false,
+          "type" : {
+            "name" : "int",
+            "bitWidth" : 32,
+            "isSigned" : true
+          },
+          "children" : [ ]
+        },{
+          "name" : "columns",
+          "nullable" : false,
+          "type" : {
+            "name" : "list"
+          },
+          "children" : [{
+            "name" : "$data$",
+            "nullable" : false,
+            "type" : {
+              "name" : "struct"
+            },
+            "children" : [{
+              "name" : "col-name",
+              "nullable" : false,
+              "type" : {
+                "name" : "utf8"
+              },
+              "children" : [ ]
+            },{
+              "name" : "root-col?",
+              "nullable" : false,
+              "type" : {
+                "name" : "bool"
+              },
+              "children" : [ ]
+            },{
+              "name" : "count",
+              "nullable" : false,
+              "type" : {
+                "name" : "int",
+                "bitWidth" : 64,
+                "isSigned" : true
+              },
+              "children" : [ ]
+            },{
+              "name" : "types",
+              "nullable" : false,
+              "type" : {
+                "name" : "struct"
+              },
+              "children" : [{
+                "name" : "timestamp-tz-micro-utc",
+                "nullable" : true,
+                "type" : {
+                  "name" : "struct"
+                },
+                "children" : [{
+                  "name" : "min",
+                  "nullable" : true,
+                  "type" : {
+                    "name" : "timestamp",
+                    "unit" : "MICROSECOND",
+                    "timezone" : "UTC"
+                  },
+                  "children" : [ ]
+                },{
+                  "name" : "max",
+                  "nullable" : true,
+                  "type" : {
+                    "name" : "timestamp",
+                    "unit" : "MICROSECOND",
+                    "timezone" : "UTC"
+                  },
+                  "children" : [ ]
+                }]
+              },{
+                "name" : "fixed-size-binary",
+                "nullable" : true,
+                "type" : {
+                  "name" : "bool"
+                },
+                "children" : [ ]
+              },{
+                "name" : "utf8",
+                "nullable" : true,
+                "type" : {
+                  "name" : "struct"
+                },
+                "children" : [{
+                  "name" : "min",
+                  "nullable" : true,
+                  "type" : {
+                    "name" : "utf8"
+                  },
+                  "children" : [ ]
+                },{
+                  "name" : "max",
+                  "nullable" : true,
+                  "type" : {
+                    "name" : "utf8"
+                  },
+                  "children" : [ ]
+                }]
+              },{
+                "name" : "set",
+                "nullable" : true,
+                "type" : {
+                  "name" : "int",
+                  "bitWidth" : 32,
+                  "isSigned" : true
+                },
+                "children" : [ ]
+              }]
+            },{
+              "name" : "bloom",
+              "nullable" : true,
+              "type" : {
+                "name" : "binary"
+              },
+              "children" : [ ]
+            }]
+          }]
+        }]
+      }]
+    }]
+  },
+  "batches" : [{
+    "count" : 1,
+    "columns" : [{
+      "name" : "nodes",
+      "count" : 1,
+      "TYPE_ID" : [3],
+      "OFFSET" : [0],
+      "children" : [{
+        "name" : "nil",
+        "count" : 0
+      },{
+        "name" : "branch-iid",
+        "count" : 0,
+        "VALIDITY" : [ ],
+        "OFFSET" : [0],
+        "children" : [{
+          "name" : "$data$",
+          "count" : 0,
+          "VALIDITY" : [ ],
+          "DATA" : [ ]
+        }]
+      },{
+        "name" : "branch-recency",
+        "count" : 0,
+        "VALIDITY" : [ ],
+        "OFFSET" : [0],
+        "children" : [{
+          "name" : "entries",
+          "count" : 0,
+          "VALIDITY" : [ ],
+          "children" : [{
+            "name" : "recency",
+            "count" : 0,
+            "VALIDITY" : [ ],
+            "DATA" : [ ]
+          },{
+            "name" : "idx",
+            "count" : 0,
+            "VALIDITY" : [ ],
+            "DATA" : [ ]
+          }]
+        }]
+      },{
+        "name" : "leaf",
+        "count" : 1,
+        "VALIDITY" : [1],
+        "children" : [{
+          "name" : "data-page-idx",
+          "count" : 1,
+          "VALIDITY" : [1],
+          "DATA" : [0]
+        },{
+          "name" : "columns",
+          "count" : 1,
+          "VALIDITY" : [1],
+          "OFFSET" : [0,5],
+          "children" : [{
+            "name" : "$data$",
+            "count" : 5,
+            "VALIDITY" : [1,1,1,1,1],
+            "children" : [{
+              "name" : "col-name",
+              "count" : 5,
+              "VALIDITY" : [1,1,1,1,1],
+              "OFFSET" : [0,14,20,25,31,38],
+              "DATA" : ["xt$system_from","xt$iid","xt$id","$data$","colours"]
+            },{
+              "name" : "root-col?",
+              "count" : 5,
+              "VALIDITY" : [1,1,1,1,1],
+              "DATA" : [1,1,1,0,1]
+            },{
+              "name" : "count",
+              "count" : 5,
+              "VALIDITY" : [1,1,1,1,1],
+              "DATA" : ["1","1","1","3","1"]
+            },{
+              "name" : "types",
+              "count" : 5,
+              "VALIDITY" : [1,1,1,1,1],
+              "children" : [{
+                "name" : "timestamp-tz-micro-utc",
+                "count" : 5,
+                "VALIDITY" : [1,0,0,0,0],
+                "children" : [{
+                  "name" : "min",
+                  "count" : 5,
+                  "VALIDITY" : [1,0,0,0,0],
+                  "DATA" : [1577836800000000,0,0,0,0]
+                },{
+                  "name" : "max",
+                  "count" : 5,
+                  "VALIDITY" : [1,0,0,0,0],
+                  "DATA" : [1577836800000000,0,0,0,0]
+                }]
+              },{
+                "name" : "fixed-size-binary",
+                "count" : 5,
+                "VALIDITY" : [0,1,0,0,0],
+                "DATA" : [0,1,0,0,0]
+              },{
+                "name" : "utf8",
+                "count" : 5,
+                "VALIDITY" : [0,0,1,1,0],
+                "children" : [{
+                  "name" : "min",
+                  "count" : 5,
+                  "VALIDITY" : [0,0,1,1,0],
+                  "OFFSET" : [0,0,0,3,7,7],
+                  "DATA" : ["","","foo","blue",""]
+                },{
+                  "name" : "max",
+                  "count" : 5,
+                  "VALIDITY" : [0,0,1,1,0],
+                  "OFFSET" : [0,0,0,3,6,6],
+                  "DATA" : ["","","foo","red",""]
+                }]
+              },{
+                "name" : "set",
+                "count" : 5,
+                "VALIDITY" : [0,0,0,0,1],
+                "DATA" : [0,0,0,0,3]
+              }]
+            },{
+              "name" : "bloom",
+              "count" : 5,
+              "VALIDITY" : [1,1,1,1,1],
+              "OFFSET" : [0,38,76,106,180,218],
+              "DATA" : ["3a30000003000000090000000c0000000f0000002000000022000000240000004c012a6e08db","3a30000003000000070000000b0000000f0000002000000022000000240000005b7f64446d09","3a300000020000000100010002000000180000001c0000007009b09df031","3a30000006000000020000000300000005000100090001000c0000000d000100380000003a0000003c0000004000000044000000460000009f70a283626c35c99d5d5189ea1cd84ef37d","3a3000000300000005000000060000000e00000020000000220000002400000050edb6bd8355"]
+            }]
+          }]
+        }]
+      }]
+    }]
+  }]
+}

